### ISSUE TITLE
Update coding standard to indicate preference for TestCase methods which being with "assert" and which do not have an underscore in their name

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2631]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2631|
|Type|enhancement|
|Created|2007-05-08 16:59:43Z|
|Branch|https://github.com/twisted/twisted/tree/test-standard-2631|

TestCase offers ```failUnless``` as well as ```assertTrue``` and a number of other similar pairings.  The coding standard should indicate that ```assertTrue``` is the preferred spelling and the ```fail``` spellings should be avoided where possible.  It should also express the preference of ```assertTrue``` over ```assert_```.

<details><summary>Searchable metadata</summary>

```
trac-id__2631 2631
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_test_standard_2631 branches/test-standard-2631
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1178643583000000 1178643583000000
changetime__1214400857000000 1214400857000000
version__None None
owner__ 
cc__ralphm
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Oh.  It should express the preference of the singular spelling over the plural spelling as well.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_2"></a>|@jml commented|
|-|-|

I think this is a step in the right direction, _but_:
 * I prefer `failUnless` over `assertTrue`.
 * Twisted is already choc-full of coding standard violations. Adding this rule will increase the number of violations. It would be good to:
   * Fix the existing violations of the existing coding standards.
   * Have an automated tool for checking standards compliance.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

I used to prefer ```failUnless``` over ```assertTrue```, but radix pointed out that ```assertTrue``` is preferable for technical reasons (it has "assert" in it, so it's easily searchable, it's consistent with the other assert functions).

I'm all for fixing things.  I'm not sure if you're suggesting a specific action in this case, though, and we shouldn't wait for the rest of the code base to catch up before applying this standard to new contributions.

I'm also all for automated tools, but I'm not going to implement it, and we shouldn't wait for someone else to before enacting this for new code.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_4"></a>|@pitrou commented|
|-|-|

My two cents... I've always found failUnless makes code less easily understandable, because it's a double negative, while assertTrue is straightforward.

(while "unless" is common is the Perl world, it is not in the Python world and for good reasons IMHO)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

I think that making the coding standard more public and accessible is an important part of this ticket.  We've been developing an informal and ad-hoc definition of "good" code, and this emerging organizational knowledge needs to be presented to people at a useful time in the process of their contribution of code, which is _before_ a review informs them of it after the fact (or, for that matter, when the reviewer informs them, a place to put a hyperlink to would be nice).

I've definitely done a few reviews which could have been much more productive if the contributor were aware of the coding standard beforehand, and I feel like the fault lies entirely with us at this point.  Although, probably nobody will read it when it is prominently linked either, at least we'll have the moral high ground :).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [22378]) Branching to 'dns-str-eq-2631'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

Just kidding. :/

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [23923]) Branching to 'dns-str-eq-2631-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [23923]) Branching to 'dns-str-eq-2631-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

Crap, again, crap.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [23985]) Branching to 'test-standard-2631'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun removed owner|
|-|-|

Some changes made.  These documents all need a lot of work. :/

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve set owner to @exarkun|
|-|-|

Nice improvements. Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set status to `closed`|
|-|-|

(In [24019]) Merge test-standard-2631

Author: exarkun
Reviewer: therve
Fixes: [#2631](https://github.com/twisted/twisted/issues/2631)

Update the testing documentation to mention recent additions to trial (addCleanup,
the -R option is implicit now), note that tests go into packages related to them,
change the sample test methods to use the test_ naming convention, and point out
that assertFoo is prefered over failUnlessFoo.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
